File: practice/InversePairs.py
# -*- coding:utf-8 -*-
class Solution:
    def __init__(self):
        self._cnt = 0

    def InversePairs(self, data):
        # write code here
        self._data = data
        self._tmp = [0] * len(data)
        self.mergeSort(0, len(data) - 1)
        return self._cnt % 1000000007

    def mergeSort(self, start, end):
        if start >= end:
            return
        mid = (end + start) // 2
        self.mergeSort(start, mid)
        self.mergeSort(mid + 1, end)
        self.merge(start, mid, end)

    def merge(self, start, mid, end):
        i, j, k = start, mid + 1, start
        while i <= mid or j <= end:
            if i > mid:
                self._tmp[k] = self._data[j]
                j += 1
            elif j > end:
                self._tmp[k] = self._data[i]
                i += 1
            elif self._data[i] <= self._data[j]:
                self._tmp[k] = self._data[i]
                i += 1
            else:
                self._tmp[k] = self._data[j]
                j += 1
                self._cnt += (mid - i + 1)
            k += 1
        for i in range(end - start + 1):
            self._data[i + start] = self._tmp[i + start]


if __name__ == '__main__':
    s = Solution()
    print(s.InversePairs([7, 5, 6, 4]))

File: practice/InversePairs.java
public class InversePairs {
    private long cnt = 0;
    private int[] tmp;

    public int InversePairs(int[] array) {
        tmp = new int[array.length];
        mergeSort(array, 0, array.length - 1);
        return (int) (cnt % 1000000007);
    }

    private void mergeSort(int[] array, int start, int end) {
        if (start >= end) {
            return;
        }
        int mid = (start + end) / 2;
        mergeSort(array, start, mid);
        mergeSort(array, mid + 1, end);
        merge(array, start, mid, end);
    }

    private void merge(int[] array, int start, int mid, int end) {
        int i = start, j = mid + 1, k = start;
        while (i <= mid || j <= end) {
            if (i > mid) {
                tmp[k] = array[j++];
            } else if (j > end) {
                tmp[k] = array[i++];
            } else if (array[i] <= array[j]) {
                tmp[k] = array[i++];
            } else {
                tmp[k] = array[j++];
                cnt += mid - i + 1;
            }
            k++;
        }
        for (k = start; k <= end; k++) {
            array[k] = tmp[k];
        }
    }
}
